From 520d3162834b5e4b37059b0d753909084f7ed299 Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@foss.st.com>
Date: Tue, 21 Jan 2025 09:34:32 +0100
Subject: [PATCH 1523/1695] drivers: stm32_cpu_opp: apply default OPP during
 suspend

Apply default OPP during
- LPLV-Stop(1-2) and Standby1 on MP2
- Standby on MP1.
Restore previous OPP at resume.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git c7c303326fe31125f7420f24cbb272aa4c44ad5a drivers: stm32_cpu_opp: apply default OPP during suspend]
Change-Id: I79555bb2d093878b3970a6bd80cf73dd770eb9b4
Signed-off-by: Pascal Paillet <p.paillet@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/430409
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Tested-by: Pascal PAILLET <p.paillet@st.com>
Reviewed-by: Pascal PAILLET <p.paillet@st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 core/drivers/stm32_cpu_opp.c | 109 +++++++++++++++++++----------------
 1 file changed, 60 insertions(+), 49 deletions(-)

diff --git a/core/drivers/stm32_cpu_opp.c b/core/drivers/stm32_cpu_opp.c
index cf80cb195..ac955818e 100644
--- a/core/drivers/stm32_cpu_opp.c
+++ b/core/drivers/stm32_cpu_opp.c
@@ -23,6 +23,7 @@
 #ifdef CFG_SCMI_SCPFW
 #include <scmi_agent_configuration.h>
 #endif
+#include <stm32mp_pm.h>
 #include <stm32_util.h>
 #include <trace.h>
 #include <util.h>
@@ -41,6 +42,7 @@ struct cpu_dvfs {
  * struct cpu_opp - CPU operating point
  *
  * @current_opp: Index of current CPU operating point in @dvfs array
+ * @default_opp: Index of OPP to set at boot or during suspend in @dvfs array
  * @opp_count: Number of cells of @dvfs
  * @clock: CPU clock handle
  * @regul: CPU regulator supply handle
@@ -49,10 +51,10 @@ struct cpu_dvfs {
  * @scp_regulator: Regulator instance exposed to scp-firmware SCMI DVFS
  * @scp_levels_desc: Description of voltage levels for scp-firmware SCMI DVFS
  * @scp_cpu_opp_levels_uv: Array of voltage levels described by @scp_levels_desc
- * @default_opp_freq: Operating point frequency to use during initialization
  */
 struct cpu_opp {
 	unsigned int current_opp;
+	unsigned int default_opp;
 	unsigned int opp_count;
 	struct clk *clock;
 	struct regulator *regul;
@@ -62,7 +64,6 @@ struct cpu_opp {
 	struct regulator scp_regulator;
 	struct regulator_voltages_desc scp_levels_desc;
 	int *scp_cpu_opp_levels_uv;
-	unsigned int default_opp_freq;
 #endif
 };
 
@@ -391,7 +392,6 @@ TEE_Result optee_scmi_server_init_dvfs(const void *fdt __unused,
 	unsigned int *dvfs_opp_mv = NULL;
 	size_t opp = 0;
 	struct cpu_dvfs *sorted_dvfs = NULL;
-	unsigned int scpfw_default_opp_index = UINT_MAX;
 
 	assert(agent_cfg && channel_cfg);
 
@@ -461,11 +461,7 @@ TEE_Result optee_scmi_server_init_dvfs(const void *fdt __unused,
 	for (opp = 0; opp < cpu_opp.opp_count; opp++) {
 		dvfs_khz[opp] = sorted_dvfs[opp].freq_khz;
 		dvfs_mv[opp] = sorted_dvfs[opp].volt_uv / U(1000);
-		if (cpu_opp.default_opp_freq == dvfs_khz[opp])
-			scpfw_default_opp_index = opp;
 	}
-	if (scpfw_default_opp_index == UINT_MAX)
-		panic();
 
 	free(sorted_dvfs);
 
@@ -495,7 +491,7 @@ TEE_Result optee_scmi_server_init_dvfs(const void *fdt __unused,
 
 	channel_cfg->perfd[0] = (struct scmi_perfd){
 		.name = "CPU DVFS",
-		.initial_opp = scpfw_default_opp_index,
+		.initial_opp = cpu_opp.default_opp,
 		.dvfs_opp_count = cpu_opp.opp_count,
 		.dvfs_opp_khz = dvfs_opp_khz,
 		.dvfs_opp_mv = dvfs_opp_mv,
@@ -558,12 +554,27 @@ TEE_Result stm32_cpu_opp_read_level(unsigned int *level)
 }
 #endif /*CFG_SCPFW_MOD_DVFS*/
 
-static TEE_Result cpu_opp_pm(enum pm_op op, unsigned int pm_hint __unused,
-			     const struct pm_callback_handle *hdl __unused)
+static TEE_Result set_opp(unsigned int opp)
 {
-	unsigned long clk_cpu = 0;
-	unsigned int opp = 0;
+	unsigned long clk_cpu = clk_get_rate(cpu_opp.clock);
+	TEE_Result res = TEE_ERROR_GENERIC;
 
+	assert(clk_cpu);
+	if (cpu_opp.dvfs[opp].freq_khz * 1000U >= clk_cpu)
+		res = set_voltage_then_clock(opp);
+	else
+		res = set_clock_then_voltage(opp);
+
+	if (res)
+		DMSG("Failed to set OPP %ukHz",
+		     cpu_opp.dvfs[opp].freq_khz);
+
+	return res;
+}
+
+static TEE_Result cpu_opp_pm(enum pm_op op, unsigned int pm_hint,
+			     const struct pm_callback_handle *hdl __unused)
+{
 	assert(op == PM_OP_SUSPEND || op == PM_OP_RESUME);
 
 	/* nothing to do if OPP is managed by Linux and not by SCMI */
@@ -571,12 +582,20 @@ static TEE_Result cpu_opp_pm(enum pm_op op, unsigned int pm_hint __unused,
 	    !IS_ENABLED(CFG_SCPFW_MOD_DVFS))
 		return TEE_SUCCESS;
 
-	/* nothing to do if RCC clock tree is not lost */
+#if defined(CFG_STM32MP25) || defined(CFG_STM32MP23) || defined(CFG_STM32MP21)
+	/* Nothing to do on Stop1, LP-Stop1 modes */
+	if (PM_HINT_PLATFORM_STATE(pm_hint) == PM_CORE_LEVEL)
+		return TEE_SUCCESS;
+#else
+	/* Nothing to do if RCC clock tree is not lost */
 	if (!PM_HINT_IS_STATE(pm_hint, CONTEXT))
 		return TEE_SUCCESS;
+#endif
 
-#ifdef CFG_SCPFW_MOD_DVFS
 	if (op == PM_OP_SUSPEND) {
+#ifdef CFG_SCPFW_MOD_DVFS
+		unsigned int opp = 0;
+
 		/*
 		 * When CFG_SCPFW_MOD_DVFS is enabled, save CPU OPP on suspend
 		 * for restoration at resume. If CPU is not in an expected
@@ -599,33 +618,35 @@ static TEE_Result cpu_opp_pm(enum pm_op op, unsigned int pm_hint __unused,
 		if (opp >= cpu_opp.opp_count) {
 			EMSG("Unexpected OPP state, select default");
 
-			for (opp = 0; opp < cpu_opp.opp_count; opp++)
-				if (cpu_opp.default_opp_freq ==
-				    cpu_opp.dvfs[opp].freq_khz)
-					break;
-
-			if (opp >= cpu_opp.opp_count)
-				panic();
+			opp = cpu_opp.default_opp;
 		}
-
-		DMSG("Suspend to OPP %u", opp);
 		cpu_opp.current_opp = opp;
+#endif
+		/*
+		 * Force default OPP before suspend
+		 * On STM32MP2, it is not allowed to enter LPLV-Stop
+		 * with overdrive voltage on VDDCPU
+		 */
+		DMSG("Suspend to OPP %u", cpu_opp.default_opp);
 
-		return TEE_SUCCESS;
+		if (cpu_opp.current_opp == cpu_opp.default_opp)
+			return TEE_SUCCESS;
+
+		return set_opp(cpu_opp.default_opp);
 	}
-#endif
 
 	if (op == PM_OP_RESUME) {
-		opp = cpu_opp.current_opp;
+		DMSG("Resume to OPP %u", cpu_opp.current_opp);
 
-		DMSG("Resume to OPP %u", opp);
+		/*
+		 * Optimize resume from Stop modes: do not reset
+		 * the OPP if we where on default before suspend.
+		 */
+		if (!PM_HINT_IS_STATE(pm_hint, CONTEXT) &&
+		    cpu_opp.current_opp == cpu_opp.default_opp)
+			return TEE_SUCCESS;
 
-		clk_cpu = clk_get_rate(cpu_opp.clock);
-		assert(clk_cpu);
-		if (cpu_opp.dvfs[opp].freq_khz * 1000U >= clk_cpu)
-			return set_voltage_then_clock(opp);
-		else
-			return set_clock_then_voltage(opp);
+		return set_opp(cpu_opp.current_opp);
 	}
 
 	return TEE_SUCCESS;
@@ -661,7 +682,6 @@ static TEE_Result stm32_cpu_opp_get_dt_subnode(const void *fdt, int node)
 	uint64_t freq_khz = 0;
 	uint64_t freq_khz_opp_def = 0;
 	uint32_t volt_uv = 0;
-	unsigned long clk_cpu = 0;
 	unsigned int i = 0;
 	int subnode = -1;
 	TEE_Result res = TEE_ERROR_GENERIC;
@@ -675,7 +695,7 @@ static TEE_Result stm32_cpu_opp_get_dt_subnode(const void *fdt, int node)
 	if (!cpu_opp.dvfs)
 		return TEE_ERROR_OUT_OF_MEMORY;
 
-	cpu_opp.current_opp = cpu_opp.opp_count;
+	cpu_opp.default_opp = cpu_opp.opp_count;
 
 	fdt_for_each_subnode(subnode, fdt, node) {
 		if (stm32_cpu_opp_is_supported(fdt, subnode) != TEE_SUCCESS)
@@ -733,7 +753,7 @@ static TEE_Result stm32_cpu_opp_get_dt_subnode(const void *fdt, int node)
 		if (fdt_getprop(fdt, subnode, "st,opp-default", NULL) &&
 		    freq_khz > freq_khz_opp_def) {
 			opp_default = true;
-			cpu_opp.current_opp = i;
+			cpu_opp.default_opp = i;
 			freq_khz_opp_def = freq_khz;
 		}
 
@@ -748,20 +768,11 @@ static TEE_Result stm32_cpu_opp_get_dt_subnode(const void *fdt, int node)
 	}
 
 	/* Select the max "st,opp-default" node as current OPP */
-#ifdef CFG_SCPFW_MOD_DVFS
-	cpu_opp.default_opp_freq = freq_khz_opp_def;
-#endif
-	clk_cpu = clk_get_rate(cpu_opp.clock);
-	assert(clk_cpu);
-	if (freq_khz_opp_def * 1000U > clk_cpu)
-		res = set_voltage_then_clock(cpu_opp.current_opp);
-	else
-		res = set_clock_then_voltage(cpu_opp.current_opp);
-
-	if (res) {
-		EMSG("Failed to set default OPP %u", cpu_opp.current_opp);
+	res = set_opp(cpu_opp.default_opp);
+	if (res)
 		goto err;
-	}
+
+	cpu_opp.current_opp = cpu_opp.default_opp;
 
 	register_pm_driver_cb(cpu_opp_pm, NULL, "cpu-opp");
 
-- 
2.47.2

